File: packages/shared/lib/services/auth_service.dart
// Flutter Auth Service
// Shared between customer_app and technician_app - Numeric Role Version

import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'dart:developer' as dev;

// Numeric Role Definitions
// 1: Admin, 2: Manager, 3: Staff, 4: Technician, 5: Customer, 0: Pending
class UserRoles {
  static const int admin = 1;
  static const int manager = 2;
  static const int staff = 3;
  static const int technician = 4;
  static const int customer = 5;
  static const int pending = 0;
}

class AuthUser {
  final String uid;
  final String email;
  final String displayName;
  final String phoneNumber;
  final int role;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? avatar;
  final bool isVerified;
  final String status;
  final String? source;
  final String? address;
  final List<String>? specializations;

  AuthUser({
    required this.uid,
    required this.email,
    required this.displayName,
    required this.phoneNumber,
    required this.role,
    required this.createdAt,
    required this.updatedAt,
    this.avatar,
    required this.isVerified,
    required this.status,
    this.source,
    this.address,
    this.specializations,
  });

  AuthUser copyWith({
    String? uid,
    String? email,
    String? displayName,
    String? phoneNumber,
    int? role,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? avatar,
    bool? isVerified,
    String? status,
    String? source,
    String? address,
    List<String>? specializations,
  }) {
    return AuthUser(
      uid: uid ?? this.uid,
      email: email ?? this.email,
      displayName: displayName ?? this.displayName,
      phoneNumber: phoneNumber ?? this.phoneNumber,
      role: role ?? this.role,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      avatar: avatar ?? this.avatar,
      isVerified: isVerified ?? this.isVerified,
      status: status ?? this.status,
      source: source ?? this.source,
      address: address ?? this.address,
      specializations: specializations ?? this.specializations,
    );
  }

  factory AuthUser.fromFirestore(DocumentSnapshot doc, User firebaseUser) {
    final data = doc.data() as Map<String, dynamic>? ?? {};
    String status = data['status'] ?? 'pending';
    bool verified = (status == 'active');

    return AuthUser(
      uid: firebaseUser.uid,
      email: firebaseUser.email ?? data['email'] ?? '',
      displayName: data['displayName'] ?? firebaseUser.displayName ?? '',
      phoneNumber: data['phoneNumber'] ?? firebaseUser.phoneNumber ?? '',
      role: data['role'] ?? UserRoles.pending,
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ??
          DateTime.parse(firebaseUser.metadata.creationTime.toString()),
      updatedAt: (data['updatedAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      avatar: data['avatar'] ?? firebaseUser.photoURL,
      isVerified: verified,
      status: status,
      source: data['source'],
      address: data['address'],
      specializations: data['specializations'] != null
          ? List<String>.from(data['specializations'])
          : null,
    );
  }
}

class AuthException implements Exception {
  final String message;
  final String? code;

  AuthException(this.message, {this.code});

  @override
  String toString() => message;
}

class AuthService {
  static final AuthService _instance = AuthService._internal();
  final _auth = FirebaseAuth.instance;
  final _firestore = FirebaseFirestore.instance;

  factory AuthService() {
    return _instance;
  }

  AuthService._internal();

  Stream<AuthUser?> get currentUserStream {
    return _auth.authStateChanges().asyncMap((firebaseUser) async {
      if (firebaseUser == null) return null;
      try {
        final userDoc = await _firestore.collection('nguoiDung').doc(firebaseUser.uid).get();
        return userDoc.exists ? AuthUser.fromFirestore(userDoc, firebaseUser) : null;
      } catch (e) {
        return null;
      }
    });
  }

  Future<AuthUser?> getCurrentUser({bool fromServer = false}) async {
    try {
      final firebaseUser = _auth.currentUser;
      if (firebaseUser == null) return null;
      final userDoc = await _firestore
          .collection('nguoiDung')
          .doc(firebaseUser.uid)
          .get(fromServer ? const GetOptions(source: Source.server) : null);
      return userDoc.exists ? AuthUser.fromFirestore(userDoc, firebaseUser) : null;
    } catch (e) {
      return null;
    }
  }

  Future<AuthUser> loginWithEmail({required String email, required String password}) async {
    try {
      await _auth.signInWithEmailAndPassword(email: email, password: password);
      final authUser = await getCurrentUser(fromServer: true);
      if (authUser == null) {
        await _auth.signOut();
        throw AuthException('Tài khoản chưa có dữ liệu trên hệ thống.');
      }
      if (authUser.status != 'active') {
        await _auth.signOut();
        throw AuthException('Tài khoản của bạn đang chờ quản trị viên phê duyệt.');
      }
      return authUser;
    } on FirebaseAuthException catch (e) {
      String message = 'Email hoặc mật khẩu không chính xác.';
      if (e.code == 'user-not-found') message = 'Email này chưa được đăng ký.';
      throw AuthException(message, code: e.code);
    } catch (e) {
      if (e is AuthException) rethrow;
      throw AuthException('Đăng nhập thất bại.');
    }
  }

  Future<void> logout() async {
    await _auth.signOut();
  }

  Future<void> sendPasswordReset(String email) async {
    try {
      final userQuery = await _firestore
          .collection('nguoiDung')
          .where('email', isEqualTo: email)
          .limit(1)
          .get();

      if (userQuery.docs.isEmpty) {
        throw AuthException('Email không tồn tại trong hệ thống.');
      }

      await _auth.sendPasswordResetEmail(email: email);
    } on FirebaseAuthException catch (e) {
      throw AuthException('Không thể gửi mail reset mật khẩu.', code: e.code);
    } catch (e) {
      if (e is AuthException) rethrow;
      throw AuthException('Lỗi hệ thống khi gửi email.');
    }
  }

  Future<void> resetPassword(String code, String newPassword) async {
    try {
      await _auth.confirmPasswordReset(code: code, newPassword: newPassword);
    } on FirebaseAuthException catch (e) {
      String message = 'Mã xác nhận không hợp lệ hoặc đã hết hạn.';
      if (e.code == 'weak-password') message = 'Mật khẩu quá yếu.';
      throw AuthException(message, code: e.code);
    } catch (e) {
      throw AuthException('Không thể đặt lại mật khẩu.');
    }
  }

  Future<AuthUser> signupWithEmail({
    required String email,
    required String password,
    required String displayName,
    required String phoneNumber,
    required int role,
    required String source,
  }) async {
    try {
      final userCredential = await _auth.createUserWithEmailAndPassword(email: email, password: password);
      await userCredential.user!.updateDisplayName(displayName);
      
      final String initialStatus = (source == 'customer_app') ? 'active' : 'pending';
      
      await _firestore.collection('nguoiDung').doc(userCredential.user!.uid).set({
        'uid': userCredential.user!.uid,
        'email': email,
        'displayName': displayName,
        'phoneNumber': phoneNumber,
        'role': role,
        'status': initialStatus,
        'source': source,
        'createdAt': FieldValue.serverTimestamp(),
        'updatedAt': FieldValue.serverTimestamp(),
      });
      
      if (initialStatus == 'pending') {
        await _auth.signOut();
        throw AuthException('Đăng ký thành công! Vui lòng chờ quản trị viên kích hoạt tài khoản.');
      }
      
      final authUser = await getCurrentUser(fromServer: true);
      return authUser ?? (throw AuthException('Lỗi đồng bộ dữ liệu sau đăng ký.'));
    } on FirebaseAuthException catch (e) {
      throw AuthException(e.code == 'email-already-in-use' ? 'Email đã được sử dụng.' : 'Đăng ký thất bại.', code: e.code);
    } catch (e) {
      if (e is AuthException) rethrow;
      throw AuthException('Lỗi trong quá trình đăng ký.');
    }
  }
}

File: apps/admin_web/src/services/authService.ts
/**
 * Firebase Authentication Service - Centralized Instance
 */

import { initializeApp, deleteApp } from 'firebase/app';
import {
  getAuth,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  sendPasswordResetEmail,
  confirmPasswordReset,
  signOut,
  updateProfile,
  deleteUser as firebaseDeleteUser,
} from 'firebase/auth';
import type { User, Auth } from 'firebase/auth';
import {
  Firestore,
  doc,
  getDoc,
  setDoc,
  serverTimestamp,
} from 'firebase/firestore';
import { auth, firestore, firebaseConfig } from '../app/firebase.config';
import {
  AuthUser,
  LoginCredentials,
  SignupCredentials,
  UserRole,
} from '../types/auth';

/**
 * Hàm hỗ trợ ép kiểu ngày tháng an toàn tuyệt đối
 */
const safeToDate = (value: any): Date => {
  if (!value) return new Date();
  if (typeof value.toDate === 'function') return value.toDate();
  if (value instanceof Date) return value;
  if (value && typeof value === 'object' && value.seconds) return new Date(value.seconds * 1000);
  const d = new Date(value);
  return isNaN(d.getTime()) ? new Date() : d;
};

export const getDb = (): Firestore => firestore;
export const getAuthInstance = (): Auth => auth;

export const convertFirebaseUser = async (user: User): Promise<AuthUser> => {
  const userDocRef = doc(firestore, 'nguoiDung', user.uid);
  const userDoc = await getDoc(userDocRef);

  if (!userDoc.exists()) {
    return {
      uid: user.uid,
      email: user.email || '',
      displayName: user.displayName || 'Người dùng mới',
      phoneNumber: '',
      role: UserRole.PENDING,
      createdAt: new Date(),
      updatedAt: new Date(),
      isVerified: false,
      status: 'pending',
      source: 'admin_web'
    };
  }

  const data = userDoc.data();
  return {
    uid: user.uid,
    email: user.email || data.email || '',
    displayName: data.displayName || user.displayName || 'Thành viên',
    phoneNumber: data.phoneNumber || '',
    // Đảm bảo role luôn là kiểu number
    role: (data.role !== undefined && data.role !== null) ? Number(data.role) : UserRole.PENDING,
    isVerified: data.status === 'active',
    status: data.status || 'pending',
    source: data.source || 'admin_web',
    createdAt: safeToDate(data.createdAt),
    updatedAt: safeToDate(data.updatedAt),
    baseSalary: data.baseSalary || 0,
    commissionPerOrder: data.commissionPerOrder || 0
  };
};

export const loginWithEmail = async (credentials: LoginCredentials): Promise<AuthUser> => {
  const userCredential = await signInWithEmailAndPassword(auth, credentials.email, credentials.password);
  const authUser = await convertFirebaseUser(userCredential.user);

  // Cho phép Admin đăng nhập ngay cả khi chưa active (để fix hệ thống) hoặc nếu status là active
  if (authUser.status !== 'active' && Number(authUser.role) !== UserRole.ADMIN) {
    await signOut(auth);
    throw new Error('Tài khoản của bạn chưa được kích hoạt hoặc đã bị khóa.');
  }

  return authUser;
};

export const signupWithEmail = async (credentials: SignupCredentials): Promise<void> => {
  let createdFirebaseUser: User | null = null;
  try {
    const userCredential = await createUserWithEmailAndPassword(auth, credentials.email, credentials.password);
    createdFirebaseUser = userCredential.user;
    await updateProfile(createdFirebaseUser, { displayName: credentials.displayName });

    const userDocRef = doc(firestore, 'nguoiDung', createdFirebaseUser.uid);
    await setDoc(userDocRef, {
      uid: createdFirebaseUser.uid,
      email: credentials.email,
      displayName: credentials.displayName,
      phoneNumber: credentials.phoneNumber,
      role: Number(credentials.role ?? UserRole.PENDING),
      status: 'pending',
      source: credentials.source || 'admin_web',
      createdAt: serverTimestamp(),
      updatedAt: serverTimestamp(),
    });
    await signOut(auth);
  } catch (error: any) {
    if (createdFirebaseUser) await firebaseDeleteUser(createdFirebaseUser).catch(() => {});
    throw error;
  }
};

/**
 * CẤP TÀI KHOẢN (ADMIN ONLY)
 */
export const adminCreateAuthUser = async (credentials: SignupCredentials): Promise<string> => {
  const tempAppName = `temp-app-${Date.now()}`;
  const tempApp = initializeApp(firebaseConfig, tempAppName);
  const tempAuth = getAuth(tempApp);

  try {
    const userCredential = await createUserWithEmailAndPassword(tempAuth, credentials.email, credentials.password);
    const uid = userCredential.user.uid;
    await signOut(tempAuth);
    await deleteApp(tempApp);
    return uid;
  } catch (error: any) {
    await deleteApp(tempApp);
    throw error;
  }
};

export const logout = async (): Promise<void> => { await signOut(auth); };

export const getCurrentUser = async (): Promise<AuthUser | null> => {
  const user = auth.currentUser;
  if (!user) return null;
  try { return await convertFirebaseUser(user); } catch { return null; }
};

export const sendPasswordReset = async (email: string) => {
  await sendPasswordResetEmail(auth, email);
};

export const resetPasswordWithCode = async (code: string, newPassword: string) => {
  await confirmPasswordReset(auth, code, newPassword);
};

export const initializeFirebase = (): void => {
  try {
    if (!auth || !firestore) throw new Error('Firebase instances not available');
  } catch (error) {
    console.error('Firebase initialization failed:', error);
  }
};

export const isFirebaseConfigured = (): boolean => {
  const requiredEnvVars = ['VITE_FIREBASE_API_KEY', 'VITE_FIREBASE_AUTH_DOMAIN', 'VITE_FIREBASE_PROJECT_ID'];
  return requiredEnvVars.every(envVar => !!import.meta.env[envVar]);
};
